dulwich/credentials.py

```diff
@@ -0,0 +1,125 @@
+# credentials.py -- support for git credential helpers
+
+# Copyright (C) 2022 Daniele Trifirò <daniele@iterative.ai>
+#
+# Dulwich is dual-licensed under the Apache License, Version 2.0 and the GNU
+# General Public License as public by the Free Software Foundation; version 2.0
+# or (at your option) any later version. You can redistribute it and/or
+# modify it under the terms of either of these two licenses.
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# You should have received a copy of the licenses; if not, see
+# <http://www.gnu.org/licenses/> for a copy of the GNU General Public License
+# and <http://www.apache.org/licenses/LICENSE-2.0> for a copy of the Apache
+# License, Version 2.0.
+#
+
+"""Support for git credential helpers
+
+https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage
+
+Currently Dulwich supports only the `get` operation
+
+"""
+import os
+import subprocess
+import sys
+from typing import Any, Dict, List, Optional, Union
+
+from dulwich.config import ConfigDict
+
+
+class CredentialNotFoundError(Exception):
+    pass
+
+
+class CredentialHelper:
+    """Helper for retrieving credentials for http/https git remotes
+
+    Usage:
+    >>> helper = CredentialHelper("store") # Use `git credential-store`
+    >>> credentials = helper.get("https://github.com/dtrifiro/aprivaterepo")
+    >>> username = credentials["username"]
+    >>> password = credentials["password"]
+    """
+
+    def __init__(self, command: str):
+        self.command = command
+        self._run_kwargs: Dict[str, Any] = {}
+        if self.command[0] == "!":
+            # On Windows this will only work in git-bash and/or WSL2
+            self._run_kwargs["shell"] = True
+
+    def _prepare_cmd(self) -> Union[str, List[str]]:
+        if self.command[0] == "!":
+            return self.command[1:]
+
+        argv = self.command.split(" ")
+        if not os.path.isabs(argv[0]):
+            cmd = ["git", f"credential-{argv[0]}"]
```

Comment:
To clarify the intention here isn't to introduce any dependency on C Git, the expected behavior for 3rd party credential helpers is to follow the executable naming conventions set by C Git:

> Generally speaking, rule (3) above is the simplest for users to specify. Authors of credential helpers should make an effort to assist their users by naming their program "git-credential-$NAME", and putting it in the $PATH or $GIT_EXEC_PATH during installation, which will allow a user to enable it with git config credential.helper $NAME.

Although that does mean we probably should just use `git-credential-{name}` instead of relying on `git credential-{name}` @dtrifiro

Reply:
I agree.

The rationale for using `git credential-foo` instead of `git-credential-foo` here was that (I'm assuming) most users will be using cgit provided credential helpers (such as `git-credential-store`, `git-credential-osxkeychain`, `git-credential-gnome-keyring`, etc), which are in `GIT_EXEC_PATH` (which is not in `PATH`). Using `git` made sure that `GIT_EXEC_PATH` was also checked.

Pushed a patch which fixes the issue: if `git-credential-foo` is not in `PATH` and `git` cli is available, `GIT_EXEC_PATH` is retrieved using `git --exec-path` and we check if `git-credential-foo` is available in `GIT_EXEC_PATH`. If `git` cli is not available, nothing is done, and calling `subprocess.run` will just raise `FileNotFoundError` (which will result in `CredentialNotFound` exception)